Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Debug.azsli

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <Atom/Features/SrgSemantics.azsli>
+
+#include <scenesrg.srgi>
+#include <viewsrg.srgi>
+
+#include <Atom/RPI/BitOperations.azsli>
+
+#ifndef ENABLE_SHADER_DEBUGGING
+#define ENABLE_SHADER_DEBUGGING     1
+#endif
+
+#if ENABLE_SHADER_DEBUGGING
+    option bool o_shader_debugging_enabled = true;
+#else
+    #define o_shader_debugging_enabled false
+#endif
+
+// Options controlling various aspects of the rendering
+enum class RenderDebugOptions
+{
+    DebugEnabled,
+
+    OverrideBaseColor,
+    OverrideRoughness,
+    OverrideMetallic,
+
+    UseDebugLight,
+
+    EnableNormalMaps,
+    EnableDetailNormalMaps,
+
+    EnableDiffuseLighting,
+    EnableSpecularLighting,
+    EnableDirectLighting,
+    EnableIndirectLighting,
+};
+
+// Specifies what debug info to render to the view
+enum class RenderDebugViewMode
+{
+    None,
+
+    Normal,
+    Tangent,
+    Bitangent,
+    
+    BaseColor,
+    Albedo,
+    Roughness,
+    Metallic,
+};
+
+bool IsDebuggingEnabled()
+{
+#if ENABLE_SHADER_DEBUGGING
+    return o_shader_debugging_enabled && SceneSrg::m_debuggingEnabled;
+#else
+    return false;
+#endif
+}
+
+// Placeholder version of the above function until o_shader_debugging_enabled can be hooked up to all shaders
+bool IsDebuggingEnabled_PLACEHOLDER()
+{
+#if ENABLE_SHADER_DEBUGGING
+    return SceneSrg::m_debuggingEnabled;
+    //return o_shader_debugging_enabled && SceneSrg::m_debuggingEnabled;
+#else
+    return false;
+#endif
+}
+
+
+uint GetDebugOptions()
+{
+    return ViewSrg::m_renderDebugOptions;
+}
+
+uint GetDebugOption(RenderDebugOptions debugOption)
+{
+    uint optionIndex = (uint)debugOption;
+    uint debugOptions = GetDebugOptions();
+    return GetBit(debugOptions, optionIndex);
+}
+
+RenderDebugViewMode GetRenderDebugViewMode()
+{
+    return (RenderDebugViewMode)ViewSrg::m_renderDebugViewMode;
+}
+
+void DebugModifyOutput(inout float4 outDiffuse, inout float4 outSpecular, inout float4 outAlbedo, inout float4 outSpecularF0,
```

Comment:
I wonder if the interface would be easier to use if instead of having a big function containing every input, and branching internally to determine what data to emit, we broke this up instead. For example:

```hlsl
void DebugNormals(inout float4 outDiffuse, inout float4 outSpecular, float3 normal);
void DebugTangents(inout float4 outDiffuse, inout float4 outSpecular, float3 tangent);
void DebugBitangents(inout float4 outDiffuse, inout float4 outSpecular, float3 bitangent);
void DebugRoughness(inout float4 outDiffuse, inout float4 outSpecular, float roughness);
```

Then, the implementation of the functions above would be no-op if the corresponding view mode did not enable it. Shaders would then call the functions that were relevant (e.g. suppose we have a custom shader, or refactor rendering so that metalness isn't available in a certain pass, or tangents are generated later, etc.).

Reply:
Feel separating it out would be more prone to error. It would be possible to forget one of the view modes and not realize it until someone runs it with the material in question with that view mode. I also means if we update the function to include more view modes and inputs, it will be easier to fix it everywhere because of compilation errors will point out all the spots very loudly. If adding a new view mode with separate functions, it's possible to miss a spot or for a customer with their own custom shader implementing debug views to sync and not know they should update their shader to support the new view mode.
I agree with you that separate calls is a bit more flexible and allows user to omit the functions that aren't applicable, like metallic on a subsurface scattering material, but on the other hand if the info is not available, it's practical for the user to be forced to think about what they should return (i.e. 0 for lack of metallic). With separate functions, omitting the missing surface param like metallic will just have the shader output diffuse when the metallic view mode is selected, which isn't the desired behavior.

Reply:
Makes sense although the converse statement is also true. A unified api means it's not easy to change the debug functionality without breaking customer code. As we add more and more parameters to debug, it's possible this function signature will fail to scale. Maybe a good compromise is a middle ground. DebugPBR and DebugGeometry etc